### ISSUE TITLE
ignore certain POJO methods

### ISSUE THREAD

**Issue by clixtec** [axelspringer/polymorphia#4]
Hi, I've just started using Polymorphia for POJO mapping with MongoDB. I'm using the Java driver 3.5.0 but have had issues getting its built-in PojoCodecProvider to behave as I need.

I have a class that is blowing up because it has a method like this:
`public <T> Map<String,T> getAttributesOfType(T argument) {
    ...
}`
(This is not an actual getter, as such.)

There is a method in ReflectionHelper being called - `private static void getMethodTypePairsRecursive(ParameterizedType type, List<MethodTypePair> currentList, Map<String, Type> realTypeMap)` - that scans all declared and inherited methods of my class and verifies that return types are fully resolved. It then blows up inside TypeUtils of commons-lang3 when it finds the second parameter not to be resolved.

The whole stack trace from MongoCollection.insertOne downwards is:
`java.lang.IllegalArgumentException: null type argument at index 1
	at org.apache.commons.lang3.Validate.noNullElements(Validate.java:513)
	at org.apache.commons.lang3.reflect.TypeUtils.parameterizeWithOwner(TypeUtils.java:1504)
	at de.bild.codec.ReflectionHelper.inferRealType(ReflectionHelper.java:106)
	at de.bild.codec.ReflectionHelper.getMethodTypePairsRecursive(ReflectionHelper.java:184)
	at de.bild.codec.ReflectionHelper.getMethodTypePairsRecursive(ReflectionHelper.java:158)
	at de.bild.codec.ReflectionHelper.getMethodTypePairsRecursive(ReflectionHelper.java:180)
	at de.bild.codec.ReflectionHelper.getMethodTypePairsRecursive(ReflectionHelper.java:158)
	at de.bild.codec.ReflectionHelper.getMethodTypePairsRecursive(ReflectionHelper.java:180)
	at de.bild.codec.ReflectionHelper.getMethodTypePairsRecursive(ReflectionHelper.java:158)
	at de.bild.codec.ReflectionHelper.getDeclaredAndInheritedMethods(ReflectionHelper.java:150)
	at de.bild.codec.BasicReflectionCodec.<init>(BasicReflectionCodec.java:64)
	at de.bild.codec.PojoContext.resolve(PojoContext.java:93)
	at de.bild.codec.PolymorphicReflectionCodec.<init>(PolymorphicReflectionCodec.java:45)
	at de.bild.codec.PojoContext.calculateCodec(PojoContext.java:156)
	at de.bild.codec.PojoContext.getCodec(PojoContext.java:66)
	at de.bild.codec.PojoContext.get(PojoContext.java:44)
	at de.bild.codec.PojoCodecProvider.get(PojoCodecProvider.java:43)
	at org.bson.codecs.configuration.ProvidersCodecRegistry.get(ProvidersCodecRegistry.java:43)
	at org.bson.codecs.configuration.ProvidersCodecRegistry.get(ProvidersCodecRegistry.java:55)
	at org.bson.codecs.configuration.ProvidersCodecRegistry.get(ProvidersCodecRegistry.java:37)
	at com.mongodb.MongoCollectionImpl.getCodec(MongoCollectionImpl.java:591)
	at com.mongodb.MongoCollectionImpl.insertOne(MongoCollectionImpl.java:314)`

Thinking that perhaps the name was confusing things, I changed it not to look like a getter, but that had no effect.

Is there a way to annotate a method as not to be mapped during encoding?

Thanks,
Sean

**Comment by webermich** [comment]
Hi Sean,

thanks for using Polymorphia.
You raise an interesting issue.

In short, I will be able to fix your problem tomorrow.

TLDR;
Right now getters and setters are not considered while scanning pojos.
Solely fields are considered for encoding/decoding. 
You can exclude fields by annotating them with @de.bild.codec.annotations.Tranisent

You cannot (at present) mark methods as being transient, as methods anyway won't be used for encoding/decoding purposes.

Lifecycle-management methods, annotated with @Postload or @PreSave will be searched while scanning the pojo. This - as seen in your case - may lead to the exeption you experience.
I just haven't thought of a generic method signature while scanning methods.

The good news is, the fix is easy, as I don't need to resolve full method signatures for methods annotated with @Postload or @PreSave. It is only needed to find these methods, but not to resolve their exact signatures.

Once more, thanks for using Polymorphia. Any feedback is appreciated.

**Comment by clixtec** [comment]
Thanks Michael. Can you clarify: does this mean I will need to annotate all my "strange" methods with @postload or @presave? Or will your method scan now only look for those annotated methods? Without a full understanding of the code, it sounds to me like the latter is the better approach.

**Comment by webermich** [comment]
Hi Sean,

you don't have to annotate your "strange" methods.
Actually Polymorphia is meant to make a developers life as easy as possible, hence you should be able to use any method etc. without special handling in your pojos.

You actually found an error that I just fixed. Please update to Polymorphia 1.6.1

Background information:
When a Pojo class is being scanned, all fields are used scanned and only fields are used for encoding to the mongo or for decoding respectivley.

Scanning of methods was introduced later and is only needed for life-cycle management.
You can annotate methods with @PreSave or @PostLoad. Before an entitity is written to the database all @PreaSave-annotated methods run. You could then prepare your entity for persistence or sent an event or alike.

You are right, it might be better to search only methods annotated with @PreSave and @PostLoad. Right now the code scans all methods within a pojos class hierarchy, resolves their full generic types and build a list via Refelctionhelper.getDeclaredAndInheritedMethods(...) 
The parameter "T" as being a TypeVariable could not be resolved. It is only know at the caller side of the method but not within the pojo class itself.
As we do not really need to resolve this type variable I made teh code more robust and you should not experience the error you saw again.

**Comment by webermich** [comment]
One more thing.
Out of curiosity. What problems did you run into using mongos PojoCodecProvider?

Thanks Michael

**Comment by clixtec** [comment]
Mainly newbie problems. It seemed to be unable to deal with certain types - I register a class explicitly, then get an exception that a codec for that class cannot be found. The tangle of CodecRegistry implementations down inside the driver is pretty wild!

Having said that, I think I have found another Polymorphia issue which I will file shortly.
